### PR TITLE
Add transform dialect op to reorder transpose

### PR DESCRIPTION
**Pull request by harsh-nod**
This patch adds a transform dialect op that moves
a transpose op before its preceding elementwise op.

### CODE REVIEW

File: compiler/src/iree/compiler/Codegen/LLVMGPU/Utils/LLVMGPULayoutAnalysisAndDistribution.cpp [hunk 1]
```diff
@@ -1041,6 +1041,38 @@ static bool isMatmulTransposeB(vector::ContractionOp contractOp) {
...
+      transposedOperands.push_back(transposed);
+    }
+    SmallVector<Type> resultTypes{transposedOperands.front().getType()};
+    Operation *newOp =
+        rewriter.create(op->getLoc(), op->getName().getIdentifier(),
+                        transposedOperands, resultTypes, op->getAttrs());
+    rewriter.replaceAllUsesWith(transposeOp.getResult(), newOp->getResult(0));
+  }
+}
+
```
Comment (ThomasRaoux) [inline compiler/src/iree/compiler/Codegen/LLVMGPU/Utils/LLVMGPULayoutAnalysisAndDistribution.cpp]: Let move that to LLVMGPUUtils.cpp

File: compiler/src/iree/compiler/Codegen/LLVMGPU/TransformExtensions/LLVMGPUExtensionsOps.td [hunk 2]
```diff
@@ -514,4 +514,43 @@ def LayoutAnalysisAndDistributionOp :
...
 }
 
+def ReorderTransposeOp :
+  Op<Transform_Dialect, "iree.reorder_transpose",
+    [FunctionalStyleTransformOpTrait,
+     MemoryEffectsOpInterface,
+     TransformEachOpTrait,
+     TransformOpInterface]> {
+  let description = [{
+    Targets the whole func op and does the following:
```
Comment (ThomasRaoux) [inline compiler/src/iree/compiler/Codegen/LLVMGPU/TransformExtensions/LLVMGPUExtensionsOps.td]: It would be nice to have a quick description in plain english

**Review by ThomasRaoux — COMMENTED** [review]
Looks good, few style related comments.

File: compiler/src/iree/compiler/Codegen/LLVMGPU/Utils/LLVMGPULayoutAnalysisAndDistribution.cpp (hunk 1 above)
Comment (harsh-nod) [reply compiler/src/iree/compiler/Codegen/LLVMGPU/Utils/LLVMGPULayoutAnalysisAndDistribution.cpp]: Sounds good, will move.

File: compiler/src/iree/compiler/Codegen/LLVMGPU/TransformExtensions/LLVMGPUExtensionsOps.td (hunk 2 above)
Comment (harsh-nod) [reply compiler/src/iree/compiler/Codegen/LLVMGPU/TransformExtensions/LLVMGPUExtensionsOps.td]: Sure will add one

**Review by ThomasRaoux — APPROVED** [review]
